### ISSUE TITLE
 adaptive adjacency matrix

### ISSUE THREAD

**Issue by WangYibucea** [rusty1s/pytorch_geometric#2141]
How to build an adaptive adjacency matrix by pyg, so as to complete the adaptive graph convolution

**Comment by rusty1s** [comment]
Can you be a bit more precise? You can generally create a fully-connected graph with edge weights set to `1`, and backprop through edge weights to find only meaningful edges.

**Comment by WangYibucea** [comment]
Such like that. The problem is that i don‘t know how to make the Adjacency Matrix (N*N) to the Edge Index (2*E), so i encounter error.




import torch
import torch.nn as nn
import torch.nn.functional as F
from torch_geometric.nn import TAGConv

class pygTAGLayer(nn.Module):
    def __init__(self, in_c, out_c, num_nodes, device):
        super(pygTAGLayer, self).__init__()
        self.TAG = TAGConv(in_c, out_c, K=3)
        self.nodevec1 = nn.Parameter(torch.randn(num_nodes, 10).to(device), requires_grad=True).to(device)
        self.nodevec2 = nn.Parameter(torch.randn(10, num_nodes).to(device), requires_grad=True).to(device)

    def forward(self, x2):
        adp = F.softmax(F.relu(torch.mm(self.nodevec1, self.nodevec2)), dim=1)
        batch_size, num_vertices, features, time_stamps = x2.shape
        gx = []
        for i in range(batch_size):
            x = x2[i, :, :, :]
            x = x.squeeze(0)
            for t in range(time_stamps):
                y = x[:, :, t]
                y = y.squeeze(-1)
                out = self.TAG(y, adp)
                gx.append(out.unsqueeze(0))
        gx = torch.cat(gx, dim=0)  # (batch*time,N,F)
        gx = gx.reshape(batch_size, time_stamps, num_vertices, -1)  # B,T,N,F
        gx = gx.permute(0, 2, 3, 1)  # B,N,F,T
        gx = F.leaky_relu(gx, negative_slope=0.1)
        return gx

x = torch.FloatTensor(32, 307, 64, 12)
model = pygTAGLayer(64, 12, 307, 'cpu')
y = model(x)

![image](https://user-images.githubusercontent.com/65903769/108748911-1a24a080-757a-11eb-99f3-643054eff610.png)

**Comment by WangYibucea** [comment]
```
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch_geometric.nn import TAGConv

class pygTAGLayer(nn.Module):
    def __init__(self, in_c, out_c, num_nodes, device):
        super(pygTAGLayer, self).__init__()
        self.TAG = TAGConv(in_c, out_c, K=3)
        self.nodevec1 = nn.Parameter(torch.randn(num_nodes, 10).to(device), requires_grad=True).to(device)
        self.nodevec2 = nn.Parameter(torch.randn(10, num_nodes).to(device), requires_grad=True).to(device)


    def forward(self, x2):
        adp = F.softmax(F.relu(torch.mm(self.nodevec1, self.nodevec2)), dim=1)
        batch_size, num_vertices, features, time_stamps = x2.shape
        gx = []
        for i in range(batch_size):
            x = x2[i, :, :, :]
            x = x.squeeze(0)
            for t in range(time_stamps):
                y = x[:, :, t]
                y = y.squeeze(-1)
                out = self.TAG(y, adp)
                gx.append(out.unsqueeze(0))
        gx = torch.cat(gx, dim=0)  # (batch*time,N,heads,F)
        gx = gx.reshape(batch_size, time_stamps, num_vertices, -1)  # B,T,N,F
        gx = gx.permute(0, 2, 3, 1)  # B,N,F,T
        gx = F.leaky_relu(gx, negative_slope=0.1)
        return gx
x = torch.FloatTensor(32, 307, 64, 12)
model = pygTAGLayer(64, 12, 307, 'cpu')
y = model(x)
```

**Comment by rusty1s** [comment]
You can convert a `N` times `N` adjacency matrix to the required format via:
```python
edge_index = adp.nonzero(as_tuple=False).t()
edge_weight = adp[edge_index[0], edge_index[1]]
```
I hope this helps!

**Comment by WangYibucea** [comment]
Thanks for your help! Great help!
